### Context

The number of restaurants in New York is increasing day by day. Lots of students and busy professionals rely on those restaurants due to their hectic lifestyles. Online food delivery service is a great option for them. It provides them with good food from their favorite restaurants. A food aggregator company FoodHub offers access to multiple restaurants through a single smartphone app.

The app allows the restaurants to receive a direct online order from a customer. The app assigns a delivery person from the company to pick up the order after it is confirmed by the restaurant. The delivery person then uses the map to reach the restaurant and waits for the food package. Once the food package is handed over to the delivery person, he/she confirms the pick-up in the app and travels to the customer's location to deliver the food. The delivery person confirms the drop-off in the app after delivering the food package to the customer. The customer can rate the order in the app. The food aggregator earns money by collecting a fixed margin of the delivery order from the restaurants.

### Objective

The food aggregator company has stored the data of the different orders made by the registered customers in their online portal. They want to analyze the data to get a fair idea about the demand of different restaurants which will help them in enhancing their customer experience. Suppose you are hired as a Data Scientist in this company and the Data Science team has shared some of the key questions that need to be answered. Perform the data analysis to find answers to these questions that will help the company to improve the business.

### Data Description

The data contains the different data related to a food order. The detailed data dictionary is given below.

### Data Dictionary

* order_id: Unique ID of the order
* customer_id: ID of the customer who ordered the food
* restaurant_name: Name of the restaurant
* cuisine_type: Cuisine ordered by the customer
* cost: Cost of the order
* day_of_the_week: Indicates whether the order is placed on a weekday or weekend (The weekday is from Monday to Friday and the weekend is Saturday and Sunday)
* rating: Rating given by the customer out of 5
* food_preparation_time: Time (in minutes) taken by the restaurant to prepare the food. This is calculated by taking the difference between the timestamps of the restaurant's order confirmation and the delivery person's pick-up confirmation.
* delivery_time: Time (in minutes) taken by the delivery person to deliver the food package. This is calculated by taking the difference between the timestamps of the delivery person's pick-up confirmation and drop-off information

### Importing the required libraries

In [16]:
# import libraries for data manipulation
import numpy as np
import pandas as pd

# import libraries for data visualization
import matplotlib.pyplot as plt
import seaborn as sns

In [17]:
# read the data
data = pd.read_csv('data/foodhub_order.csv')
# Copying data to another variable to avoid any changes to the original data
df = data.copy()
# returns the first 5 rows
df.head()

order_id  customer_id            restaurant_name cuisine_type  \
0   1477147       337525                    Hangawi       Korean   
1   1477685       358141  Blue Ribbon Sushi Izakaya     Japanese   
2   1477070        66393                Cafe Habana      Mexican   
3   1477334       106968  Blue Ribbon Fried Chicken     American   
4   1478249        76942           Dirty Bird to Go     American   

   cost_of_the_order day_of_the_week     rating  food_preparation_time  \
0              30.75         Weekend  Not given                     25   
1              12.08         Weekend  Not given                     25   
2              12.23         Weekday          5                     23   
3              29.20         Weekend          3                     25   
4              11.59         Weekday          4                     25   

   delivery_time  
0             20  
1             23  
2             28  
3             15  
4             24

In [18]:
# Printing number of count of each unique value in customer_id
print(df['customer_id'].value_counts())
print('-' * 50)

customer_id
52832     13
47440     10
83287      9
250494     8
259341     7
          ..
143926     1
89574      1
157711     1
66393      1
358141     1
Name: count, Length: 1200, dtype: int64
--------------------------------------------------


#### Observations:

The DataFrame has 9 columns as mentioned in the Data Dictionary. Data in each row corresponds to the order placed by a customer.

The rating has a non numerical value: 'Not given'.

The column day_of_the_week is a categorical variable.

The column customer_id in this context represents a categorical variable.
The above output shows which of the customer could be a potencial loyal client of the application.

The column order_id doesn't add valuable information for the problem context

In [19]:
df.shape

(1898, 9)

The dataframe contains 1898 rows and 9 columns

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1898 entries, 0 to 1897
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               1898 non-null   int64  
 1   customer_id            1898 non-null   int64  
 2   restaurant_name        1898 non-null   object 
 3   cuisine_type           1898 non-null   object 
 4   cost_of_the_order      1898 non-null   float64
 5   day_of_the_week        1898 non-null   object 
 6   rating                 1898 non-null   object 
 7   food_preparation_time  1898 non-null   int64  
 8   delivery_time          1898 non-null   int64  
dtypes: float64(1), int64(4), object(4)
memory usage: 133.6+ KB


#### Observations:

* All columns have 1898 observations. This indicates that there aren't 'Null' values presents.
* The columns order_id, customer_id, food_preparation_time and delivery time are integers data types.
* The columns restaurant_name, cuisine_type, day_of_the_week and rating are object data type.
* The column cost_of_the_order is a float data type.

In [21]:
df.isnull().sum()

order_id                 0
customer_id              0
restaurant_name          0
cuisine_type             0
cost_of_the_order        0
day_of_the_week          0
rating                   0
food_preparation_time    0
delivery_time            0
dtype: int64

* There are no missing values in the data.

In [22]:
# Since order_id doesn't add valuable information it will be droped
df = df.drop('order_id', axis=1)
#Calculate the statistical summary for numerical columns
df.describe().T

count           mean            std      min  \
customer_id            1898.0  171168.478398  113698.139743  1311.00   
cost_of_the_order      1898.0      16.498851       7.483812     4.47   
food_preparation_time  1898.0      27.371970       4.632481    20.00   
delivery_time          1898.0      24.161749       4.972637    15.00   

                            25%        50%          75%        max  
customer_id            77787.75  128600.00  270525.0000  405334.00  
cost_of_the_order         12.08      14.14      22.2975      35.41  
food_preparation_time     23.00      27.00      31.0000      35.00  
delivery_time             20.00      25.00      28.0000      33.00

In [23]:
df.describe(exclude = 'number').T

count unique          top  freq
restaurant_name  1898    178  Shake Shack   219
cuisine_type     1898     14     American   584
day_of_the_week  1898      2      Weekend  1351
rating           1898      4    Not given   736

#### Observations:


*   The minimun time that takes for food to be prepared is 20 min
*   The average time that takes for food to be prepared is 27 min
*   The maximun time that takes for food to be prepared is 35 min

In [24]:
order_no_rating = df[df['rating'] == 'Not given']
print(order_no_rating['rating'].value_counts())
print('-' * 50)

rating
Not given    736
Name: count, dtype: int64
--------------------------------------------------


#### Observations:
There are 736 order not rated from a total of 1898 orders. That is a 38.8% of the orders in the sample are not rated.